cxx4opencl/diff2openclc.txt

```diff
@@ -0,0 +1,120 @@
+
+In this section some of the most common areas where it is
+expected to behave differently are described.
+
+==== Implicit Conversions
+
+{cpp} is much more strict about conversions between types
+especially those that are performed implicitly by the compiler.
+For example it is not possible to convert non-`const` object
+to `const` explicitly. For details please refer to {cpp}17
+(`ISO/IEC 14882:2017`) `[conv]`.
+
+[source,c]
+----------
+void foo(){
+  const int* ptrconst;
+  int* ptr = ptrconst; // rejected in C++ for OpenCL
+}
+----------
+
+The same situation applies to narrowing conversions in
+initialization lists ({cpp}17 `[dcl.init.list]`). 
+
+[source,c]
+----------
+struct mytype {
+ int i;
+};
+void foo(uint par){
+  mytype var = {
+    .i = par
+  };
+}
+----------
+
+Note that some compilers will allow to silence this error
+using some flags (e. g. in Clang `-Wno-error=c++11-narrowing`
+can be used).
+
+Among other common conversions that won't be compiled in C++
+mode there are pointer to integers or integer to pointer
+conversions.
+
+[source,c]
+----------
+void foo(){
+  int* ptr;
+  int i = ptr;
+}
+----------
+
+==== Null Literal
+
+In C and OpenCL C null literal is being defined using other
+language features as it is not represented explicitly i.e.
+commonly it is defined as
+
+[source,c]
+----------
+#define NULL (void*)0
+----------
+
+In {cpp} there is an explicit builtin literal `nullptr`
+that should be used instead ({cpp}17 `[lex.nullptr]`).
+
+{cpp} for OpenCL does not define `NULL` and therefore any
+source code using it must be modified to use `nullptr`
+instead. However as a workaround to avoid large modifications
+`NULL` can also be defined/aliased to `nullptr` in custom
+headers or using command line flag `-D`. It is not recommended
+to reuse C definition of `NULL` in {cpp} for OpenCL as it may
+cause compilation failures in cases that work for C.
+
+[source,c]
+----------
+void foo(){
+ int* ptr = NULL; // inivalid initialization of int* with void*
+}
+----------
+
+==== Use of restrict
+
+C++17 does not support `restrict` and therefore {cpp} for OpenCL
+can not support it either. Some compilers might provide extensions
+with some functionality of `restrict` in {cpp} e. g. `__restrict`
+in Clang.
+
+==== Limitations of goto statements
+
+{cpp} is more restrictive with respect to entering the scope of
+variables than C. It is not possible to jump forward over a variable
+declaration statement apart from some exceptions detailed in {cpp}17
+`[stmt.dcl]`.
+
+[source,c]
+----------
+if (cond)
+  goto label;
+int n = foo();
+label:  // invalid: jumping forward over declaration of n
+   // ...
+----------
+
+==== Use of Clang Blocks
+
+Clang Blocks that are defined by ObjC language are not supported
+and their use can be replaced by lambdas ({cpp}17 `[expr.prim.lambda]`).
+
+This means that builtin function `enqueue_kernel` with OpenCL C
+prototype `OpenCL v2.0 s6.13.17` is not supported in {cpp} for OpenCL.
```

Comment:
There are other BIF that rely on blocks, possibly in extensions. I would suggest generalising this statement a bit, maybe as follows: 

```suggestion
This means that builtin functions using blocks, such as `enqueue_kernel`,
are not supported in {cpp} for OpenCL.
```

Reply:
Just FYI I would still like to enumerate those from the core spec here.